File: OlderCodes/Ccodes/exp-good.c
/* 
************************************************************************
*  exp-good.c: A good algorithm for calculating exponential            *
*                                                                      *
*  From: "A SURVEY OF COMPUTATIONAL PHYSICS" 
   by RH Landau, MJ Paez, and CC BORDEIANU 
   Copyright Princeton University Press, Princeton, 2008.
   Electronic Materials copyright: R Landau, Oregon State Univ, 2008;
   MJ Paez, Univ Antioquia, 2008; & CC BORDEIANU, Univ Bucharest, 2008
   Support by National Science Foundation                             
*                                                                      *
*  related programs: exp-bad.c.c                                       *
************************************************************************
*/

#include <stdio.h>
#include <math.h>

#define min 1E-10                           /* limit for accuracy */
#define max 10.                             /* maximum for x */
#define step 0.1                            /* interval */

main ()
{
   double x, sum, element;         
   int n;          

   FILE *output;               /* save results in */
   output=fopen("exp-good.dat", "w");      /* exp-good.dat */
   
   for (x=0.0; x<=max; x+=step)
   {
      sum = element = 1.;                  /* reset variables  */
      n   = 0;
      do                                   /* sum terms until */
      {                                    /* accuracy is reached */
         n++; 
         element *= -x/n;                  /* calculate next element */
         sum += element;
      }while ((sum == 0) || (fabs(element/sum) > min));
      
      fprintf(output, "%f\t%e\n", x, sum);
   }
   printf("results saved in exp-good.dat\n");     
   fclose(output);
}     

File: OlderCodes/Fortran77codes/exp-good.f
c  From: "A SURVEY OF COMPUTATIONAL PHYSICS" 
c  by RH Landau, MJ Paez, and CC BORDEIANU 
c  Copyright Princeton University Press, Princeton, 2008.
c  Electronic Materials copyright: R Landau, Oregon State Univ, 2008;
c  MJ Paez, Univ Antioquia, 2008; and CC BORDEIANU, Univ Bucharest, 2008.
c  Support by National Science Foundation                           
c	                                                                 
c  exp-good.f:  good algorithm for calculating exponential
c               related programs: exp-bad.f

      Program expgood
			
 	      Implicit none
c                             limit for accuracy, max in x, step in x
       	Real*8 element, min, max, step, sum, x
	      Integer i, j, n
				
	      min = 1E-10
	      max = 10.0
	      step = 0.1
	      Open(6, File = 'exp-good.dat', Status = 'Unknown')
c                                                           execution
	      Do 10 x = 0, max, step
	        sum = 1
	        element = 1
	        Do 20 n = 1, 10000
	          element = element*(-x)/n
	          sum = sum + element
	    if ((abs(element/sum) .lt. min) .AND. (sum .ne. 0)) then 
	            Write (6,*) x, sum
c                       N.B. since no "while" in f77, need use "goto"
	            GoTo 10
	          Endif       
 20	      Continue
 10 	  Continue
        Close(6)
        Stop 'data saved in exp-good.dat'
      End
       
